### PR TITLE
Fix `Count_Distinct` in Snowflake

### LINKED ISSUE #10611: Fix `Count_Distinct` on a Boolean column in Snowflake
Currently, when trying to perform `Count_Distinct` `aggregate` on a `Boolean` column in the Snowflake backend, it fails with:
```
An unexpected dataflow error (There was an SQL error: Failed to cast variant value {"enso-null-replacement-marker":"084954a3-9f58-4a2d-aa63-3f56bc270800"} to BOOLEAN. [Query was: SELECT COUNT("Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Concatenate strs") AS "Concatenate strs", COUNT("Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Sum ints") AS "Sum ints", COUNT("Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Count Distinct bools") AS "Count Distinct bools" FROM (SELECT "Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Concatenate strs" AS "Concatenate strs", "Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Sum ints" AS "Sum ints", "Tinfo-31365771-714d-4377-bb92-da160dccbf19"."Cou (...) "strs") END, ?) || (?) AS "Concatenate strs", SUM("Tinfo-31365771-714d-4377-bb92-da160dccbf19"."ints") AS "Sum ints", COUNT(DISTINCT COALESCE("Tinfo-31365771-714d-4377-bb92-da160dccbf19"."bools", {'enso-null-replacement-marker':'084954a3-9f58-4a2d-aa63-3f56bc270800'}::variant)) AS "Count Distinct bools" FROM "Tinfo-31365771-714d-4377-bb92-da160dccbf19" AS "Tinfo-31365771-714d-4377-bb92-da160dccbf19") AS "Tinfo-31365771-714d-4377-bb92-da160dccbf19") AS "Tinfo-31365771-714d-4377-bb92-da160dccbf19"]) has been matched (at C:\NBO\enso\test\Snowflake_Tests\src\Snowflake_Spec.enso:189:13-109).
```

Apparently the 'special null replacement value' that we use does not play well with the Boolean type.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #10611 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by radeusgd** [comment]
[Extra Nightly Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml)

**Comment by radeusgd** [comment]
I wanted to see if the COUNT DISTINCT hacks we do affects performance of the query.
I tried the following code:
```
from Standard.Base import all
import Standard.Base.Errors.Illegal_State.Illegal_State

from Standard.Table import all
from Standard.Database import all
from Standard.Snowflake import all

my_snowflake =
    account_name = Environment.get "ENSO_SNOWFLAKE_ACCOUNT"
    if account_name.is_nothing then Nothing else
        get_var name =
            Environment.get name if_missing=(Panic.throw (Illegal_State.Error "ENSO_SNOWFLAKE_ACCOUNT is set, but "+name+" is not. Please set all required environment variables."))
        user = get_var "ENSO_SNOWFLAKE_USER"
        password = get_var "ENSO_SNOWFLAKE_PASSWORD"
        database = get_var "ENSO_SNOWFLAKE_DATABASE"
        schema = Environment.get "ENSO_SNOWFLAKE_SCHEMA" if_missing="PUBLIC"
        warehouse = Environment.get "ENSO_SNOWFLAKE_WAREHOUSE" if_missing=""

        resolved_password = if password.starts_with "enso://" then Enso_Secret.get password else password
        credentials = Credentials.Username_And_Password user resolved_password
        Snowflake_Details.Snowflake account_name credentials database schema warehouse

main =
    c = Database.connect my_snowflake
    n = 50000
    in_memory_table = Table.new [["X", (0.up_to n . map v-> if v%5 == 1 then Nothing else (v%5))]]
    in_memory_table.column_info.print
    in_memory_table.print
    table_name = "My_Big_Table"
    IO.println <| Duration.time_execution <|
        Panic.rethrow <| in_memory_table.select_into_database_table c table_name primary_key=[] temporary=True
    
    runs = 20
    measure_query query = 
        IO.println "Running "+query
        results = 0.up_to runs . map _->
            r = Duration.time_execution <|
                Panic.rethrow <| c.read (..Raw_SQL query)
            r.second.print
            r.first
        results.each t-> IO.println t.to_display_text

    IO.println "SIMPLE"
    measure_query 'SELECT COUNT(DISTINCT "X") FROM "'+table_name+'"'

    rnd = Random.uuid
    IO.println "old replaced"
    measure_query 'SELECT COUNT(DISTINCT COALESCE("X", {\'enso-null-replacement-marker\':\''+rnd+'\'}::variant)) FROM "'+table_name+'"'

    IO.println "new replaced+casted"
    measure_query 'SELECT COUNT(DISTINCT COALESCE("X"::variant, {\'enso-null-replacement-marker\':\''+rnd+'\'}::variant)) FROM "'+table_name+'"'
```

The biggest problem was that uploading only 50k rows took me... 5 minutes! And that's far too little rows to show any performance difference:
image

File: test/Snowflake_Tests/src/Snowflake_Spec.enso [hunk 1]
```diff
@@ -394,7 +394,7 @@ snowflake_specific_spec suite_builder default_connection db_name setup =
             table_name = Name_Generator.random_name "TimestampTZ"
             table = default_connection.get.create_table table_name [Column_Description.Value "A" (Value_Type.Date_Time with_timezone=True), Column_Description.Value "rowid" Value_Type.Integer] primary_key=[]
             table.should_succeed
-            Panic.with_finalizer default_connection.get.drop_table table.name <|
+            Panic.with_finalizer (default_connection.get.drop_table table.name) <|
```
Comment (GregoryTravis) [inline test/Snowflake_Tests/src/Snowflake_Spec.enso:397]: Does this not give a compile-time error, having too many arguments?

**Review by GregoryTravis — APPROVED** [review]

File: test/Snowflake_Tests/src/Snowflake_Spec.enso (hunk 1 above)
Comment (radeusgd) [reply test/Snowflake_Tests/src/Snowflake_Spec.enso:397]: No, such stuff is only currently resolved at runtime.

Once #9812 is implemented, we'd get a warning if static type analysis were enabled.

**Comment by radeusgd** [comment]
I've redone the analysis by manually uploading a 10M row dataset to Snowflake through the web panel and then running the queries on it. Here we can see a difference:
image

The 'Simple' flavour is `COUNT(DISTINCT X)` without the `COALESCE`, `Old` is what we had right now and `New` is the one with the fix for boolean columns. I don't see any significant difference between `Old` and `New`, so I think the fix is totally OK.

We can see however that we do pay a price for the `COALESCE` workaround that gives us the correct null behaviour. Thankfully, if we are doing distinct by a single column with `ignore_nothing=True` (non-default setting), we will get the efficient `Simple` query. Unfortunately, with the default `ignore_nothing=True` OR if we have more than one column, we will get the less efficient one.

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**